### PR TITLE
Add a service to handle migrations applies and database seeding

### PR DESCRIPTION
**Pull request by maxime-poulain**
Currently the mechanism to apply migrations and seed the database is not efficient, it is extensions methods at the Web layer.
I introduced a service that can execute logic upon launching of the application.
The implementation runs asynchronously to speed up start.

### CODE REVIEW

File: src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        if (!completedWithSuccess)
+        {
+            _logger.LogError("Seeding [{database}] database couldn't complete", currentConnection.Database);
+        }
+    }
+
+    private async Task<bool> SafeApplyMigrationsWithRetriesAsync(CatalogContext catalogContext, IDbConnection currentConnection)
+    {
+        for (var i = 0; i < 150; i++)
```
[the remark this answers is not in the document]
Comment (maxime-poulain) [reply src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs:46]: Actually that's something i wanted to get a feedback on: I have no clue what are average measure in the real world.
However we might add some kind of 10 sec of delay before even starting the process based on what you said above.

File: src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        return false;
+    }
+
+    private async Task<bool> SafeApplyMigrationsAsync(CatalogContext context)
+    {
+        try
+        {
+            await context.Database.MigrateAsync();
+            await SeedTrainersAsync(context);
```
[the remark this answers is not in the document]
Comment (maxime-poulain) [reply src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs:65]: The naming of the SafeApplyMigrationsWithRetriesAsync is indeed incorrect as it should clearly state it seeds also the data.

I might either rename the method or just pull the seed off into another method.
I

File: src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maxime-poulain) [reply src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs:46]: Well you can do delayed retires computed by some arithmetic function without installing Polly or Hangfire or whatsoever.
I didn't wanted to install Polly just for one case but, for more complex cases such http calls we might want to use a good library.

I added this and it will do the job for now:

```csharp
int DelayToWaitBetweenRetriesInMilliseconds(int retryAttempt) => (int) (Math.Max(5 - retryAttempt, 0) + Math.Pow(2, Math.Min(retryAttempt, 5))) * 1_000;
```
```
6, 6, 7, 10, 17, 32
```
32 seconds is the max value computed by the function, hence the uses of Min and Max methods.

EF will try for 20~30 seconds to open an active connection on the database.
I doubt that the SQL server instance would not be up by then. Therefore the delay could even be removed.

I chose to add this implementation to show how important is asynchronous implementations when it comes to I/O operations.
It is really important (and we can say mandatory) to use it when it comes to that kind of operations.

By switching to tasks and replace the 
```csharp
Thread.Sleep(xxxx);
```
with 
```csharp
await Tas.Delay(xxx);
```
it doesn't lock the main thread and therefore stops the startup of the API application.
It increases the availability of the API.

However there is another important question that remains:
**what do to is the migration just did fail?**
I would just go with throwing an exception and just stop the execution of the application.

File: src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxime-poulain) [reply src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs:67]: I decided to keep the the execution order, so seeding does not happen if migrations fail.
To make it more clear I renamed the method.

**Comment by maxime-poulain** [comment]
Thx @BomUbik for feedback.

File: src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BomUbik) [reply src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs]: If the migration fail, I would consider that the database is not in an appropriate state and the application should not run. If the application expects tables, columns or even data that are not there, it would most likely crash anyway. I think it's better to have it crash at startup than having it running and crash when users are on it.

[the remark this answers is not in the document]
Comment (maxime-poulain) [reply src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs]: I have added the following:
` throw new Exception("Applying migrations and seed the database was not successful, aborting the startup of the application");`

Since the encountered exception are already logged, I don't think we need a particular exception to stop the startup of the app.

File: src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    private Task SeedTrainersAsync(CatalogContext catalogContext)
+    {
+        var trainer = new Trainer(Name.Create("Victor", "vD").Value,
+            TrainerIdentity.Create("1", ApplicationType.Default).Value,
+            "Developer",
+            "Hello I am Victor van Duynen",
+            Language.Create("FR").Value);
+        catalogContext.Trainers.Add(trainer);
+        return catalogContext.SaveChangesAsync();
+    }
```
Comment (DexSmart) [inline src/Smart.FA.Catalog.Infrastructure/Services/BootStrapService.cs:96]: Should this really be merged ?

**Review by DexSmart — APPROVED** [review]

**Review by LievenPraet — APPROVED** [review]

**Review by Victor-van-Duijnen — APPROVED** [review]
